### DISCUSSION TITLE
Problem when dealing with dates

### DISCUSSION THREAD

**Discussion by csmashe** [microsoft/semantic-kernel#8921] [Q&A] [answered]
I am having a heck of a time getting Semantic Kernel (or my code) to work the same way every time when dealing with dates. 

I have a chat bot that looks up data from my database and returns answers to questions that are asked.   Its a react website using endpoints so we send a copy of the history to the endpoint every time to keep the history of the chat available for the bot. 

Endpoint code. 
`using FastEndpoints;
using HealthLog.Server.RequestModels;
using HealthLog.Server.ResponseModels;
using HealthLog.Server.Validation;
using Markdig;
using Microsoft.AspNetCore.Http.HttpResults;
using Microsoft.SemanticKernel;
using Microsoft.SemanticKernel.ChatCompletion;
using Microsoft.SemanticKernel.Connectors.OpenAI;

namespace HealthLog.Server.EndPoints.LLM;

// ReSharper disable once UnusedMember.Global
public class Chatbot(Kernel kernel) : Endpoint<ChatbotRequestModel, Results<Ok<ChatbotResponseModel>, BadRequest>>
{
    public override void Configure()
    {
        Post("/LLM/Chatbot");
        AllowAnonymous();
    }

    public override async Task<Results<Ok<ChatbotResponseModel>, BadRequest>>
        ExecuteAsync(ChatbotRequestModel req, CancellationToken ct)
    {
        var validator = new LlmChatbotValidation();
        var validationResult = await validator.ValidateAsync(req, ct);
        if (!validationResult.IsValid)
        {
            return TypedResults.Ok(new ChatbotResponseModel
            {
                Answer = Markdown.ToHtml("No Question Sent")
            });
        }

        var chat = kernel.GetRequiredService<IChatCompletionService>();

        var history = new ChatHistory();

        var isDaylightSavings = TimeZoneInfo.Local.IsDaylightSavingTime(DateTime.Now);
            var localTimeZone = isDaylightSavings ? TimeZoneInfo.Local.DaylightName : TimeZoneInfo.Local.StandardName;
        history.AddSystemMessage($"""
                                  Date/Times returned by functions are going to be in UTC. Convert all times to {localTimeZone} in your final responses to the user.
                                  If anything is marked omit still use the information to provide a response.
                                  Today is {DateTime.Now.ToLongDateString()}.
                                  Always set the time to 12:00 AM if you are only interested in the date and the time is not specified in the request.
                                  """
        );

        // Use the incoming history to add context to the current conversation
        foreach (var msg in req.History)
        {
            if (msg.Sender == "User")
            {
                history.AddUserMessage(msg.Text);
            }
            else if (msg.Sender == "Bot")
            {
                history.AddAssistantMessage(msg.Text);
            }
        }

        var message = new ChatMessageContentItemCollection
        {
            new TextContent(req.Question)
        };

        history.AddUserMessage(message);

        var executionSettings = new OpenAIPromptExecutionSettings
        {
            ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions
        };

        var result = await chat.GetChatMessageContentAsync(history, executionSettings, kernel, ct);

        return TypedResults.Ok(new ChatbotResponseModel
        {
            Answer = Markdown.ToHtml(result.Content ?? "No response from OpenAI")
        });
    }
}`

When the bot opens if i ask it did i drink coffee today if calls function get drinks and passes the date time in utc to the function. When it gets to the function its in local (which is strange) so we convert it back to utc and everything is good. 

Function GetDrinks invoking.
trce: Microsoft.SemanticKernel.KernelFunction[0]
      Function arguments: {"startDate":"2024-09-20T00:00:00Z","endDate":"2024-09-20T23:59:59Z"}
      
Function Code Here
`public class DrinkFunctions(Func<IDrinksService> drinkServiceFactory)
{
    [KernelFunction]
    [Description("Gets The Drinks That Was Drank Between 2 Dates.")]
    public async Task<List<DrinksDrank>> GetDrinks([Description("Start Date")] DateTime startDate,
        [Description("End Date")] DateTime endDate)
    {
        if (startDate > endDate)
        {
            throw new ArgumentException("Start Date cannot be greater than End Date.");
        }

        if (startDate.Kind != DateTimeKind.Utc)
        {
            startDate = startDate.ToUniversalTime();
        }

        if (endDate.Kind != DateTimeKind.Utc)
        {
            endDate = endDate.ToUniversalTime();
        }

        var drinksService = drinkServiceFactory.Invoke();
        var response = await drinksService.GetByDateRangeAsync(startDate, endDate, CancellationToken.None);
        return response;
    }
}`

This returns what is expected. 

The issue is if I ask What about yesterday it changes the call to the function to no longer be dealing with the full date and changes the time.

Microsoft.SemanticKernel.KernelFunction[0]
      Function arguments: {"startDate":"2024-09-18T04:00:00Z","endDate":"2024-09-19T03:59:59Z"}
      
      I asked it a second time What about yesterday and this time it returned the expected time frame
      Microsoft.SemanticKernel.Connectors.OpenAI.OpenAIChatCompletionService[0]
      Function call requests: DrinkFunctions-GetDrinks({"startDate":"2024-09-19T00:00:00Z","endDate":"2024-09-19T23:59:59Z"})

How can i get it to do the same thing each time? 

Note: I send the current date as part of the system message every time because I had a get date function and sometimes it would call it and other times i would guess and think it was next month or 2023.

**Comment by evchaki** [comment, marked as answer]
You can use a function call that gets the current time/date and pass that over to the LLM for function calling vs doing this in a system message. And maybe even a second one for +/- days.  You can also pass JSON to the model in the format that you want on the way out ( in the Kernel function description).
See if that will work for you.

**Reply by csmashe** [reply]
I started with that and even though it was giving back the right date sometime it was not using what was returned and going off the rails even more so i changed it to be in the system message. 

Thanks for looking at this. Sorry for the late reply. I was out of town.
